aea/aea.py

```diff
@@ -0,0 +1,290 @@
+        """
+        self._protocols = {}  # type: Dict[ProtocolId, Protocol]
+        self._handlers = {}  # type: Dict[ProtocolId, AbstractHandler]
+
+    def populate(self, directory: str) -> None:
+        """
+        Load the handlers as specified in the config and apply consistency checks.
+
+        :param directory: the filepath to the agent's resource directory.
+        :return: None
+        """
+
+        protocols_spec = importlib.util.find_spec(".".join([directory, "protocols"]))
+        protocols_packages = list(filter(lambda x: not x.startswith("__"), protocols_spec.loader.contents()))
+        logger.debug("Processing the following protocol packages: {}".format(protocols_packages))
+        for protocol_name in protocols_packages:
+            try:
+                self._add_protocol(directory, protocol_name)
+            except Exception:
+                logger.exception("Not able to add protocol {}.".format(protocol_name))
+
+    def fetch_protocol(self, envelope: Envelope) -> Optional[Protocol]:
+        """
+        Fetch the protocol for the envelope.
+
+        :pass envelope: the envelope
+        :return: the protocol id or None if the protocol is not registered
+        """
+        return self._protocols.get(envelope.protocol_id, None)
+
+    def fetch_handler(self, envelope: Envelope) -> Optional[AbstractHandler]:
+        """
+        Fetch the handler for the envelope.
+
+        :pass envelope: the envelope
+        :return: the protocol id or None if the protocol is not registered
+        """
+        return self._protocols.get(envelope.protocol_id, None)
+
+    def teardown(self) -> None:
+        """
+        Teardown the registry.
+
+        :return: None
+        """
+        self._protocols = {}
+
+    def _add_protocol(self, directory: str, protocol_name: str):
+        """
+        Add a protocol.
+
+        :param directory: the agent's resources directory.
+        :param protocol_name: the name of the protocol to be added.
+        :return: None
+        """
+        # get the serializer
+        serialization_module = importlib.import_module(".".join([directory, "protocols", protocol_name, "serialization"]))
+        classes = inspect.getmembers(serialization_module, inspect.isclass)
+        serializer_classes = list(filter(lambda x: re.match("\\w+Serializer", x[0]), classes))
+        serializer_class = serializer_classes[0][1]
+
+        logger.debug("Found serializer class {serializer_class} for protocol {protocol_name}"
+                     .format(serializer_class=serializer_class, protocol_name=protocol_name))
+        serializer = serializer_class()
+
+        # get the handler
+        handler_module = importlib.import_module(".".join([directory, "protocols", protocol_name, "handlers"]))
+        classes = inspect.getmembers(handler_module, inspect.isclass)
+        handler_classes = list(filter(lambda x: re.match("Handler", x[0]), classes))
+        handler_class = handler_classes[0][1]
```

Comment:
I think the handler should be separate from the protocol. Consider this: I add tac protocol but rather than handling it with the default tac handler I do it with some other handler I have written myself. Note, that I consider 'handlers' to do deal with 'reactions'.

Reply:
agreed.